Source/NonlinearSolvers/JacobianFunctionMF.H

```diff
@@ -0,0 +1,235 @@
+    {
+        a_U.setVal(a_val);
+    }
+
+    inline
+    RT dotProduct( const T& a_X, const T& a_Y )
+    {
+        return( a_X.dotProduct(a_Y) );
+    }
+
+    inline
+    RT norm2( const T& a_U )
+    {
+        return ( a_U.norm2() );
+    }
+
+    [[nodiscard]] inline
+    bool isDefined() const { return m_is_defined;  }
+
+    inline
+    void setBaseSolution ( const T&  a_U )
+    {
+        m_Y0.Copy(a_U);
+        m_normY0 = norm2(m_Y0);
+    }
+
+    inline
+    void setBaseRHS ( const T&  a_R )
+    {
+        m_R0.Copy(a_R);
+    }
+
+    inline
+    void setJFNKEps ( RT a_eps )
+    {
+        m_epsJFNK = a_eps;
+    }
+
+    inline
+    void setIsLinear ( bool a_isLinear )
+    {
+        m_is_linear = a_isLinear;
+    }
+
+    inline
+    void curTime ( RT a_time )
+    {
+        m_cur_time = a_time;
+    }
+
+    inline
+    void curTimeStep ( RT a_dt )
+    {
+        m_dt = a_dt;
+    }
+
+    void define( const T&, Ops* );
+
+    private:
+
+    bool m_is_defined = false;
+    bool m_is_linear = false;
+    bool m_usePreCond = false;
+    RT m_epsJFNK = RT(1.0e-6);
+    RT m_normY0;
+    RT m_cur_time, m_dt;
+    std::string m_pc_type;
+
+    T m_Z, m_Y0, m_R0, m_R;
+    Ops* m_ops;
+
+};
+
+template <class T, class Ops>
+void JacobianFunctionMF<T,Ops>::define ( const T&  a_U,
+                                         Ops*      a_ops )
+{
+    m_Z.Define(a_U);
+    m_Y0.Define(a_U);
+    m_R0.Define(a_U);
+    m_R.Define(a_U);
+
+    m_ops = a_ops;
+
+    m_is_defined = true;
+}
+
+template <class T, class Ops>
+auto JacobianFunctionMF<T,Ops>::makeVecRHS () const -> T
+{
+    T Vec;
+    Vec.Define(m_R);
+    return Vec;
+}
+
+template <class T, class Ops>
+auto JacobianFunctionMF<T,Ops>::makeVecLHS () const -> T
+{
+    T Vec;
+    Vec.Define(m_R);
+    return Vec;
+}
+
+template <class T, class Ops>
+void JacobianFunctionMF<T,Ops>::apply (T& a_dF, const T& a_dU)
+{
+    BL_PROFILE("JacobianFunctionMF::apply()");
+    using namespace amrex::literals;
+    RT const normY = norm2(a_dU); // always 1 when called from GMRES
+
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+        isDefined(),
+        "JacobianFunction::apply() called on undefined JacobianFunction");
+
+    if (normY < 1.0e-15) { a_dF.zero(); }
+    else {
+
+        RT eps;
+        if (m_is_linear) {
+            eps = 1.0_rt;
+        } else {
+            /* eps = error_rel * sqrt(1 + ||U||) / ||a||
+             * M. Pernice and H. F. Walker, "NITSOL: A Newton Iterative Solver for
+             * Nonlinear Systems", SIAM J. Sci. Stat. Comput., 1998, vol 19,
+             * pp. 302--318. */
+            if (m_normY0==0.0) { eps = m_epsJFNK*norm2(m_R0); }
+            else {
+                // m_epsJFNK * sqrt(1.0 + m_normY0) / normY
+                // above commonly used form not recommend for poorly scaled Y0
+                eps = m_epsJFNK * m_normY0 / normY;
+            }
+        }
+        const RT eps_inv = 1.0_rt/eps;
+
+        m_Z = m_Y0 + eps*a_dU;
```

Comment:
@WeiqunZhang Pointed out that this would have performance issue, since it allocates temporary `MultiFabs` under the hood. It would be better to explicitly write this with `MultiFab` operation (e.g. linear combination).

We might even remove the overloaded `+` and `*` operator for the vector of `MultiFab`.

Reply:
What would be the best way to do the following operations:

1) a_U = a_b + m_R;
2) a_F = a_U - m_R - a_b;
3) a_dF = a_dU - eps_inv*(m_R - m_R0);